### PR TITLE
rdar://problem/27134604 Swift 3 preview 5: Turn on `AnyHashable`

### PR DESCRIPTION
**Pull request by jckarter**
Clang importer support for `AnyHashable`.

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@gribozavr Mind reviewing this for beta 5?

File: lib/ClangImporter/ImportType.cpp [hunk 1]
```diff
@@ -908,12 +908,18 @@ namespace {
...
             if (unboundDecl == Impl.SwiftContext.getDictionaryDecl() ||
                 unboundDecl == Impl.SwiftContext.getSetDecl()) {
               auto &keyType = importedTypeArgs[0];
-              if (!Impl.matchesNSObjectBound(keyType))
-                keyType = Impl.getNSObjectType();
+              if (!Impl.matchesNSObjectBound(keyType)) {
+                if (auto anyHashable = Impl.SwiftContext.getAnyHashableDecl())
+                  keyType = anyHashable->getDeclaredType();
+                else
+                  keyType = Type();
```
Comment (gribozavr) [inline lib/ClangImporter/ImportType.cpp]: `ErrorType`?  If `AnyHashable` is not available, the library is broken...

File: stdlib/public/SDK/Foundation/Foundation.swift [hunk 2]
```diff
@@ -429,6 +432,9 @@ extension NSNumber
 }
 
 extension NSNumber : _HasCustomAnyHashableRepresentation {
+  // Must be @nonobjc to prevent infinite recursion trying to bridge
+  // AnyHashable to NSObject.
```
Comment (gribozavr) [inline stdlib/public/SDK/Foundation/Foundation.swift]: Does the same `@nonobjc` caveat apply to other Foundation value types?

File: stdlib/public/SDK/Foundation/Foundation.swift [hunk 3]
```diff
@@ -1231,7 +1237,8 @@ extension NSSet {
   ///   receiver.
   @objc(_swiftInitWithSet_NSSet:)
   public convenience init(set anSet: NSSet) {
```
Comment (gribozavr) [inline stdlib/public/SDK/Foundation/Foundation.swift]: This function looks like dead code to me...

Comment (gribozavr) [inline stdlib/public/SDK/Foundation/Foundation.swift]: In fact, all `*swiftInitWith*` functions do.

File: stdlib/public/SDK/Foundation/Foundation.swift [hunk 4]
```diff
@@ -1492,3 +1501,39 @@ typealias KeyedUnarchiver = NSKeyedUnarchiver
 @available(*, deprecated, renamed:"NSKeyedArchiver", message: "Please use NSKeyedArchiver")
 typealias KeyedArchiver = NSKeyedArchiver
 
+//===----------------------------------------------------------------------===//
+// AnyHashable
+//===----------------------------------------------------------------------===//
+
+extension AnyHashable : _ObjectiveCBridgeable {
```
Comment (gribozavr) [inline stdlib/public/SDK/Foundation/Foundation.swift]: Would be great to have tests for this.

File: stdlib/public/SDK/Foundation/NSStringAPI.swift [hunk 5]
```diff
@@ -1180,7 +1180,8 @@ extension String {
   /// values found in the `String`.
   public
   func propertyListFromStringsFileFormat() -> [String : String] {
-    return _ns.propertyListFromStringsFileFormat() as! [String : String]
+    return _ns.propertyListFromStringsFileFormat()! as [NSObject : AnyObject]
+      as! [String : String]
```
Comment (gribozavr) [inline stdlib/public/SDK/Foundation/NSStringAPI.swift]: For my own education, why can't we go to `[String : String]` directly?

File: lib/ClangImporter/ImportType.cpp (hunk 1 above)
Comment (jckarter) [inline lib/ClangImporter/ImportType.cpp]: Maybe. I was following what `getNSObjectType()` would have done if `NSObject` were missing.

File: stdlib/public/SDK/Foundation/NSStringAPI.swift (hunk 5 above)
Comment (jckarter) [inline stdlib/public/SDK/Foundation/NSStringAPI.swift]: It's unfortunate. We allow bridging conversions from ObjC <-> Swift, but not a Swift -> ObjC -> Swift round trip cast, which would be necessary to go from `AnyHashable -> NSObject -> String`.

File: stdlib/public/SDK/Foundation/Foundation.swift (hunk 3 above)
Comment (jckarter) [inline stdlib/public/SDK/Foundation/Foundation.swift]: Do we not want to keep these initializers around for Cocoa interop?

File: stdlib/public/SDK/Foundation/Foundation.swift (hunk 2 above)
Comment (jckarter) [inline stdlib/public/SDK/Foundation/Foundation.swift]: Probably. It was hitting `NSString` specifically, but @nonobjc doesn't hurt for the others either.

File: stdlib/public/SDK/Foundation/Foundation.swift (hunk 3 above)
Comment (gribozavr) [inline stdlib/public/SDK/Foundation/Foundation.swift]: Oh, I see, they have non-underscored Swift names.

These are strange initializers though.  They initialize NSArray from NSArray, NSString from NSString etc.  `@objc` names are also ugly, so maybe they could benefit from `@nonobjc`.  Anyway, this seems like a separate issue.

File: stdlib/public/SDK/Foundation/Foundation.swift (hunk 2 above)
Comment (gribozavr) [inline stdlib/public/SDK/Foundation/Foundation.swift]: Could you add those attributes?

File: stdlib/public/SDK/Foundation/NSStringAPI.swift (hunk 5 above)
Comment (gribozavr) [inline stdlib/public/SDK/Foundation/NSStringAPI.swift]: I see.  Yes, bridging conversions throw in a lot of unprincipled weirdness in the type system.

**Comment by gribozavr** [comment]
LGTM with the `@nonobjc` point addressed.

File: stdlib/public/SDK/Foundation/Foundation.swift (hunk 3 above)
Comment (jckarter) [inline stdlib/public/SDK/Foundation/Foundation.swift]: I suppose `copy()` covers the use case of getting a known-immutable instance just as well, but yeah, we can discuss removing these initializers separately.

File: stdlib/public/SDK/CoreImage/CoreImage.swift [hunk 6]
```diff
@@ -50,9 +50,9 @@ extension CIFilter {
 extension CISampler {
   // - (id)initWithImage:(CIImage *)im keysAndValues:key0, ...;
   convenience init(im: CIImage, elements: (String, Any)...) {
-    var dict: [NSObject : Any] = [:]
+    var dict: [AnyHashable : Any] = [:]
```
Comment (gribozavr) [inline stdlib/public/SDK/CoreImage/CoreImage.swift]: While you are changing this code, could you reserve capacity in the new dictionary?

File: stdlib/public/SDK/Foundation/NSError.swift [hunk 7]
```diff
@@ -389,11 +389,11 @@ public protocol _BridgedStoredNSError :
 }
 
 /// TODO: Better way to do this?
-internal func _stringDictToNSObjectDict(_ input: [String : Any])
-    -> [NSObject : Any] {
-  var result: [NSObject : Any] = [:]
+internal func _stringDictToAnyHashableDict(_ input: [String : Any])
+    -> [AnyHashable : Any] {
+  var result: [AnyHashable : Any] = [:]
```
Comment (gribozavr) [inline stdlib/public/SDK/Foundation/NSError.swift]: While you are changing this code, could you reserve capacity in the new dictionary?

**Comment by jckarter** [comment]
@gribozavr Added missing `@nonobjc`s in 177a029.

**Comment by gribozavr** [comment]
Sorry, I also added comments about reserving capacity.  LGTM with those.

**Pull request merged**